Feature/hackathon/docker

roryclaydon1994 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature.
     For details, see: https://github.blog/2019-02-14-introducing-draft-pull-requests/

     For a timely review/response, avoid force-pushing additional
     commits if your PR has already received reviews or comments.

     Before submitting, ensure you've done the following:
     - 📖 Read the [EUCLID RAG Contributing Guide](CONTRIBUTING.md)
     - 📖 Read the [EUCLID RAG Code of Conduct](CODE_OF_CONDUCT.md)
     - 👷‍♀️ Create small PRs, if possible.
     - ✅ Provide necessary tests for your changes.
     - 📝 Write descriptive commit messages.
     - 📗 Update related documentation and include relevant screenshots, where applicable.
     - 🧹 Ensure code follows the project's style and linting rules.

     NOTE: PRs from forked repositories need a EUCLID RAG Team member's review
     before CI builds will run. Once approved with a `/ci` reply, subsequent builds
     may proceed without manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ x] CI / DevOps
- [ ] Security

## Description

_Provide a clear and concise description of what this PR does. Explain the problem and the solution._

This PR adds container support and fixes several bugs which prevented testing of this. The changes will also be necessary if we are to release this package on `PYPI`.

- Moved `static` to `python/euclid/static` so that the static directory can be packaged in the recommended way.
- Dynamically find `static` path using `importlib` and propagate that to relevant modules
- Updated container to split the entrypoint to a new script so that we can ensure that the `ollama` server is running before launching the app. We can split the `ollama` server to a separate container and allow these to communicate in future if that is preferred. We would then need a `docker-compose.yml` to coordinate. When running locally, make sure to mount your `~/.ollama` directory so that local models are already available in the container
- Removed code which runs the model from within the script as no longer necessary

## Evidence of testing

Docker container runs:

![Screenshot from 2025-06-07 12-43-45](https://github.com/user-attachments/assets/32516683-8ddc-4d4f-80df-7af9ff53560f)

Expected output:

```

 *  Executing task: docker run --rm -it -p 8501:8501/tcp euclid_rag:latest 

curl: (7) Failed to connect to localhost port 11434 after 0 ms: Couldn't connect to server
Waiting for Ollama server to start...

Collecting usage statistics. To deactivate, set browser.gatherUsageStats to false.

WARNING: Static file serving is enabled, but no static folder found at /app/python/euclid/rag/static. To disable static file serving, set server.enableStaticServing to false.

  You can now view your Streamlit app in your browser.

  Local URL: http://localhost:8501/gai-helpers
  Network URL: http://172.17.0.7:8501/gai-helpers
  External URL: http://90.251.186.186:8501/gai-helpers
```

Log files:
- [Uploading ollama_run.log…]()
- [ollama_serve.log](https://github.com/user-attachments/files/20638649/ollama_serve.log)

## Related Tickets & Documents

<!--
For pull requests related to issues, include them below.
Follow [GitHub's guidance on linking issues to PRs](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

E.g., "closes #1234" will link the PR to issue 1234 and close the issue on merge.
-->

- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings

_Replace this line with testing instructions, any specific devices/browsers tested, and include images for UI changes._

### UI Accessibility Checklist
_If your PR includes UI changes, use this checklist:_
- [ ] Landing page loads in a reasonable time
- [ ] Screen reader compatible
- [ ] Elements have appropriate ARIA labels

## Added/Updated Tests?

- [ ] Yes
- [x ] No, and here's why: the main test is seeing if the docker container runs
- [ ] I need help with writing tests

## [Optional] Are there any post-deployment tasks we need to perform?

## [Optional] Feedback on the Development Process

_If applicable, share insights or feedback about the development process or challenges faced._

## [Optional] What GIF best describes this PR or how it makes you feel?

![my_feelings](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExOXZ5cmxzbmkxcnU5Yml2NmhnN3Zyam5uM2NxcGIxdjdvM2lpaTRvNiZlcD12MV9naWZzX3NlYXJjaCZjdD1n/3oEjHGr1Fhz0kyv8Ig/giphy.gif)

jeipollack commented:
Hi @roryclaydon1994, thanks for this PR. It's nearly midnight so I am just taking a quick skim before I look in detail at more regular hours.  I just want to comment that this PR performs changes recommended in issue #21 regarding setting up the ollama server into another module, albeit differently than proposed.  So, it would be good if issue #21 were referenced and some feedback, if applicable.

roryclaydon1994 commented:
Hi @jeipollack ,
Issue #21 still gives responsibility to the app to launch the server, which IMHO should not be its responsibility. The app should run the app and connect to an available endpoint, in this case the ollama server. The solution here is to allow the container itself to orchestrate launching the server, ensuring it is running and the requested model is pulled.
I think this should make it more obvious how the ollama server is set up and run.

jeipollack commented:
> Hi @jeipollack ,
> 
> Issue #21 still gives responsibility to the app to launch the server, which IMHO should not be its responsibility. The app should run the app and connect to an available endpoint, in this case the ollama server. The solution here is to allow the container itself to orchestrate launching the server, ensuring it is running and the requested model is pulled.
> 
> I think this should make it more obvious how the ollama server is set up and run.

Thanks for feedback, @roryclaydon1994. I hope I imply that I disagreed with the direction taken here. For record-keeping/issue-tracking I just would like it captured that relocation of the ollama server can happen via docker entrypoint.sh. Would you be able to reference your PR comment there? The issue description could be updated.

Rolgim reviewed (commented):
Nice job! It may not be the most exciting code to write, but it's solid and will be useful throughout the project. I've spotted a few things to tweak here and there, but nothing major. The code is ready to use (I've tested it successfully btw). Thanks for making the effort it'll make our lives easier later on😄

roryclaydon1994 commented on .dockerignore:
```diff
@@ -0,0 +1,21 @@
+*.yaml
+*.md
+COPYRIGHT
+LICENSE
```
Thanks, added to container now

roryclaydon1994 commented on entrypoint.sh:
```diff
@@ -0,0 +1,32 @@
+with open("python/euclid/rag/app_config.yaml", "r") as file:
+    data = yaml.safe_load(file)
+    print(data["llm"]["model"])
+')
```
Done (mostly).

I've allowed python to exit with it's standard error, then added a check to see if the llm_model variable has been correctly populated.

roryclaydon1994 commented on entrypoint.sh:
```diff
@@ -0,0 +1,32 @@
+
+
+# Wait until the server is running
+until check_server; do
```
Added in now

roryclaydon1994 commented on Dockerfile:
```diff
@@ -1,17 +1,31 @@
 # Install any needed packages specified in requirements.txt
-RUN pip install --no-cache-dir -r requirements.txt
+RUN uv pip install --system --upgrade pip
+RUN uv pip install --system -r pyproject.toml
```
Done

roryclaydon1994 replied:
Issue was ordering of layers in the `Dockerfile`. I'll push that soon

roryclaydon1994 replied:
If I change the ordering, automatic resolution of  `pyproject.toml` is possible, but then we lose the ability to cache the `pip install` which can take ages. Not an issue in prod, but for local development it's painful. A solution could be to directly mount the pip cache, but at this point I think just specifying `pyproject.toml` is the cleanest solution.
What would you like me to do?

Rolgim replied:
I agree that keeping the install step cacheable is more important for local dev where rebuild speed matters. If the current setup with uv pip install --system -r pyproject.toml is working reliably and avoids resolution issues, then it’s totally fine to leave it as it is.
Mounting the pip cache could be a nice optimization later, but it's not critical right now. So let’s keep your current approach

roryclaydon1994 commented:
## Action Complete

- Accepted all @Rolgim comments and have implemented them accordingly. Thank you for the thorough review.
- Added issue as requested by @jeipollack 

I'm doing one last round of local tests. Please let me know if you're happy with how I've implemented the changes

jeipollack commented on Dockerfile:
```diff
@@ -14,7 +14,7 @@ WORKDIR /app
 COPY pyproject.toml /app/
 
 # Install any needed packages specified in requirements.txt
```
shouldn't this comment be updated if `pyproject.toml` is replacing `requirements.txt`?

roryclaydon1994 replied:
Updated :)

SaamieVincken commented on pyproject.toml:
Just a heads up - I've also changed this list in my PR and tried to make it a bit more structrured per package type.

roryclaydon1994 replied:
Thanks Saamie. I had a quick look at our diff and most looks like it can be automerged. I think `uv` can automatically update the pyproject.toml dependencies which we should look into later :)

SaamieVincken reviewed (approved):
Hi Rory, 

PR looks great and not much to add since Jennifer and Roland both reviewed this already. Specifically thanks to approaching the issue of moving the initiation of the LLM out of chatbot.py which helps us make deployment across environments much more stable.
